### DISCUSSION TITLE
MIDI side-chains for MIDI plugins

### DISCUSSION THREAD

**Discussion by Qf4ctor** [free-audio/clap#239] [Ideas]
Hi Berlin ;)
I just wanted to chime in because I think this is a great initiative. Props to u-he and the Bitwig teams to make this happen.
I'm particularly interested in the possibility of developing MIDI plugins which is highly problematic in VST. I think this could clearly set CLAP apart and make it appealing to a lot of plugin developers. In particular I'm thinking that if it would be amazing if it could include MIDI chaining. The applications of it would be endless: a plugin could use chord MIDI from another track or vice-versa use the melody from another track to create chords from it, you could sync with a bass track to create arpegios, sync the bass with the kick of the drum track, etc etc.

**Comment by robbert-vdh** [comment]
Plugins can already have as many note ports as they want: https://github.com/free-audio/clap/blob/main/include/clap/ext/note-ports.h

**Comment by Qf4ctor** [comment]
Yes, I see that the technical part is already in place but unless the hosts are asked to allow users to connect the MIDI ports of plugins living in different tracks or to the host MIDI tracks themselves, it is unusable in practice. Maybe I should reformulate this request as being able to send info in real time from one CLAP plugin to another.

**Comment by baconpaul** [comment]
First of all, hi from NYC! It's not all berlin :)

I think what you are asking is: could there be a host which if a plugin has multiple midi ins and outs allows flexible routing of those ports right?

From an API perspective a plugin and a host can do that today. But you are right I don't think any hosts let you.

But just want to make sure I understand!

**Comment by Qf4ctor** [comment]
Hi NYC too :)

That's right, at the end of the day it's up to the host to implement the features of a plugin standard but I wonder if there would be a way to encourage them to allow users to send the output MIDI of the host tracks to the MIDI inputs of a CLAP plugin. It would be very useful if the host would tell the CLAP plugin for instance the name of the track that is being sent to a particular MIDI input.

**Reply by baconpaul** [reply]
yeah bitwig (or I guess logic with the environment if logic supported clap) would be obvious daws to do this in, as would bespoke if it became a clap host. Interesting.

**Reply by Trinitou** [reply]
In Bitwig there are two paradigms for getting additional note data into devices ATM.

- One is: you can send from note tracks to a plugin on a specific MIDI channel (always on note input port 1)
  - main use case are multitimbral instrument plugins I think
  - here the host just offers a way to the user to send notes to the main note input port
  - this concept could easily expanded by the hosts to allow the user to send to specific plugin note in ports, to specific MIDI channels
  - this should already be possible with the current note-port extension
- The second is: some Bitwigs internal note devices have the feature to select a note source from anywhere in the project's signal paths.
  - The most prominent example would be the Note Receiver device or the Harmonize device (where a second track can be used as a live harmony source)
  - Here the device needs information about the available host's note sources.
  - Something like this isn't possible yet via CLAP. Imagine a plugin in which you can select any host track (or even other plugin's note outputs) as a note sidechain source
  - Something like that could already work with the existing note-ports extension if the host lists the additional note ports with the plugin and offers a selector for the note source. (like Bitwig's audio sidechain dropdown menu for plugins but for notes as well)
    - This way the host could take care itself to prevent event feedback loops

I think it would be hard to encourage hosts from the CLAP standard side. Maybe the fact that multiple note in ports are possible should already be motivational enough? ;)
